pkg/field/field.go

```diff
@@ -186,90 +198,430 @@ func (f *Field) CalcAreaPoint(teamID int) int {
 	return Sum
 }
 
-// ActAgentsWithSaving はエージェントの行動を指定し、フィールドを変更します。ついでに保存します。
-func (f *Field) ActAgentsWithSaving(IsValid []bool, updateActions []*apispec.UpdateAction) {
+// MakeUpdateAction2s は updateActions と updateActionIDs をまとめて返します
+func (f *Field) MakeUpdateAction2s(updateActions []*apispec.UpdateAction, updateActionIDs []int) []*UpdateAction2 {
+	updateAction2s := make([]*UpdateAction2, len(updateActions))
+	for i := range updateActions {
+		updateAction2s[i] = &UpdateAction2{
+			TeamID:       updateActionIDs[i],
+			UpdateAction: updateActions[i],
+		}
+	}
+	return updateAction2s
+}
 
-	// 行動を精査します
-	// もうやったのでIsValidは信用していいデータらしい。
+// RecordCellSelectedAgents は各セルを行動先に選んでいるような行動情報の要素番号を記録します
+func (f *Field) RecordCellSelectedAgents(isValid []bool, updateActions []*apispec.UpdateAction) [][][]int {
+
+	selectedAgents := make([][][]int, f.Height)
+	for i := range selectedAgents {
+		selectedAgents[i] = make([][]int, f.Width)
+	}
 
-	// セルが行動先に選ばれた回数をカウントします
-	var DistinationCount [][]int
 	for i, updateAction := range updateActions {
-		if IsValid[i] == true {
-			x := f.Agents[updateAction.AgentID].X + updateAction.DX
-			y := f.Agents[updateAction.AgentID].Y + updateAction.DY
-			DistinationCount[x][y]++
+		if isValid[i] == true {
+			var x, y int
+			if updateActions[i].Type == "put" {
+				x = updateAction.X
+				y = updateAction.Y
+			} else {
+				x = f.Agents[updateAction.AgentID].X + updateAction.DX
+				y = f.Agents[updateAction.AgentID].Y + updateAction.DY
+			}
+			selectedAgents[y][x] = append(selectedAgents[y][x], i)
 		}
+
 	}
-	// セルに保存された回数が1回なら、実行できます(Apply:1)
-	Apply := 1
-	for i, updateAction := range updateActions {
-		if IsValid[i] == false {
-			var slise []AgentActionHistory
-			f.ActionHistories[i].AgentActionHistories = append(f.ActionHistories[i].AgentActionHistories, slise...) // sliseの中身を追加する
-			Apply = -1
-			// f.Turn
-			continue
+
+	return selectedAgents
+}
+
+// GiveNewStack は競合しているマスが入れられたスタックのポインタを返します
+func (f *Field) GiveNewStack(selectedAgentsIndex [][][]int) *stack.Stack {
+	stk := stack.New()
+	for y := range selectedAgentsIndex {
+		for x := range selectedAgentsIndex[y] {
+			if len(selectedAgentsIndex[y][x]) >= 2 {
+				for i := range selectedAgentsIndex[y][x] {
+					stk.Push(selectedAgentsIndex[y][x][i])
+				}
+			}
 		}
-		x := f.Agents[updateAction.AgentID].X + updateAction.DX
-		y := f.Agents[updateAction.AgentID].Y + updateAction.DY
-		if DistinationCount[x][y] == 1 {
-			// 動ける、動かす
-			Apply = 1
-		} else {
-			// 競合して動けない、stayにする
-			Apply = 0
+	}
+	return stk
+}
+
+// DetermineIfApplied は 行動情報が競合か許容か不正かを判定して isApply に保存します
+func (f *Field) DetermineIfApplied(isValid []bool, updateActions []*apispec.UpdateAction, selectedAgentsIndex [][][]int) []int {
+	// isApply を初期化
+	isApply := make([]int, len(updateActions))
```

Comment:
適用できるかどうか、という意味の方がいいかもしれないです
```suggestion
	isApplicable := make([]int, len(updateActions))
```